src/org/exist/storage/NodePath2.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * eXist Open Source Native XML Database
+ * Copyright (C) 2001-2018 The eXist Project
+ * http://exist-db.org
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package org.exist.storage;
+
+import org.exist.dom.persistent.IStoredNode;
+import org.w3c.dom.Element;
+
+/**
+ * @author Stanislav Jordanov <stenlee@gmail.com>
+ * @version 1.0
+ *
+ * This is an extension of class NodePath, that keeps track of a *real* node/element path, not just a QName path
+ * as its base class -- NodePath does.
+ * This is required in order to implement the feature requested/discussed here:
```

Comment:
I am a little confused by your terminology here.

What do you man by a "real node/element path"?

I think you need to use the terms that we would expect to find in XQuery. For example do you actually mean an XPath Expression? Predicates, etc. etc.?